Add-Migration generate migration with missing "using" statement when a property is an enum defined in another namespace and using a conversion.

If your model has a property of type enum and that enum is defined in another namespace than your model. When generating a migration, by default the snapshot and migration.designer.cs file will use "int" as the variable type and it works.

But when you add a `HasConversion()` statement (even if empty), the generated snapshot and designer.cs files of the migration will use the enum type. Like so:

```C#
                    b.Property<SomeEnum>("Type")
                        .HasColumnType("int");
```

This results in a project that does not build, because the generated migration files are missing the "using" statement of the enumeration namespace on top of their files.

### Include your code

Full reproducible project (just replace the connection string and table name):

```C#
using EfCoreMigrationEnum.Enums;
using Microsoft.EntityFrameworkCore;

namespace EfCoreMigrationEnum
{
    internal class Program
    {
        static async Task Main(string[] args)
        {
            using (var context = new BloggingContext())
            {
                var posts = await context.Posts.Where(t => t.Type == SomeEnum.Value2).ToListAsync();
            }
        }
    }

    public class BloggingContext : DbContext
    {
        public DbSet<Post> Posts { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder options)
        {
            options.UseSqlServer(***);
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Post>(entity =>
            {
                entity.ToTable("Post");
                entity.HasKey(x => x.ID);
                entity.Property(x => x.Type).HasConversion(e => e, e => e);
            });
        }
    }

    public class Post
    {
        public int ID { get; set; }
        public SomeEnum Type { get; set; }
    }
}

namespace EfCoreMigrationEnum.Enums
{
    public enum SomeEnum
    {
        Value1,
        Value2,
        Value3
    }
}
```

### Include verbose output

```
PM> Add-Migration MInitial -Context BloggingContext
Build started...
Build succeeded.
```

Build afterward:

```
1>C:\Users\***\source\repos\EfCoreMigrationEnum\Migrations\BloggingContextModelSnapshot.cs(32,32,32,40): error CS0246: The type or namespace name 'SomeEnum' could not be found (are you missing a using directive or an assembly reference?)
1>C:\Users\***\source\repos\EfCoreMigrationEnum\Migrations\20231207203356_MInitial.Designer.cs(35,32,35,40): error CS0246: The type or namespace name 'SomeEnum' could not be found (are you missing a using directive or an assembly reference?)
```

### Include provider and version information

EF Core version: 8.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 8.0
Operating system: Windows 11 x64
IDE: Visual Studio 2022 17.9

--- a/src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs
+++ b/src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs
@@ -527,8 +527,7 @@ public class CSharpSnapshotGenerator : ICSharpSnapshotGenerator
     }
 
     private ValueConverter? FindValueConverter(IProperty property)
-        => property.GetValueConverter()
-            ?? property.GetTypeMapping().Converter;
+        => property.GetTypeMapping().Converter;
 
     /// <summary>
     ///     Generates code for <see cref="IComplexProperty" /> objects.

--- a/src/EFCore.Design/Migrations/Design/MigrationsCodeGenerator.cs
+++ b/src/EFCore.Design/Migrations/Design/MigrationsCodeGenerator.cs
@@ -223,8 +223,7 @@ public abstract class MigrationsCodeGenerator : IMigrationsCodeGenerator
                 .SelectMany(a => GetProviderType(a.Annotatable, a.Annotation.Value!.GetType()).GetNamespaces()));
 
     private ValueConverter? FindValueConverter(IProperty property)
-        => (property.FindTypeMapping()
-            ?? Dependencies.RelationalTypeMappingSource.FindMapping(property))?.Converter;
+        => property.GetTypeMapping().Converter;
 
     private Type GetProviderType(IAnnotatable annotatable, Type valueType)
         => annotatable is IProperty property

--- a/src/EFCore.Design/Migrations/Design/MigrationsCodeGeneratorDependencies.cs
+++ b/src/EFCore.Design/Migrations/Design/MigrationsCodeGeneratorDependencies.cs
@@ -41,13 +41,16 @@ public sealed record MigrationsCodeGeneratorDependencies
         IRelationalTypeMappingSource relationalTypeMappingSource,
         IAnnotationCodeGenerator annotationCodeGenerator)
     {
+#pragma warning disable CS0612 // Type or member is obsolete
         RelationalTypeMappingSource = relationalTypeMappingSource;
+#pragma warning restore CS0612 // Type or member is obsolete
         AnnotationCodeGenerator = annotationCodeGenerator;
     }
 
     /// <summary>
     ///     The type mapper.
     /// </summary>
+    [Obsolete]
     public IRelationalTypeMappingSource RelationalTypeMappingSource { get; init; }
 
     /// <summary>

--- a/test/EFCore.Design.Tests/Migrations/Design/CSharpMigrationsGeneratorTest.ModelSnapshot.cs
+++ b/test/EFCore.Design.Tests/Migrations/Design/CSharpMigrationsGeneratorTest.ModelSnapshot.cs
@@ -140,7 +140,6 @@ namespace MyNamespace
                 eb.Property(e => e.Decimal).HasDefaultValue(decimal.MinValue);
                 eb.Property(e => e.Double).HasDefaultValue(double.MinValue); //double.NegativeInfinity
                 eb.Property(e => e.Enum).HasDefaultValue(Enum1.Default);
-                eb.Property(e => e.NullableEnum).HasDefaultValue(Enum1.Default).HasConversion<string>();
                 eb.Property(e => e.Guid).HasDefaultValue(Guid.NewGuid());
                 eb.Property(e => e.Int16).HasDefaultValue(short.MaxValue);
                 eb.Property(e => e.Int32).HasDefaultValue(int.MaxValue);
@@ -5300,7 +5299,8 @@ namespace RootNamespace
     [ConditionalFact]
     public virtual void Property_of_nullable_enum()
         => Test(
-            builder => builder.Entity<EntityWithNullableEnumType>().Property(e => e.Day),
+            builder => builder.Entity<EntityWithNullableEnumType>().Property(e => e.Day)
+                .HasConversion(d => d, d => d),
             AddBoilerPlate(
                 GetHeading()
                 + """

--- a/test/EFCore.SqlServer.HierarchyId.Tests/TestModels/Migrations/AnonymousArraySeedContext.cs
+++ b/test/EFCore.SqlServer.HierarchyId.Tests/TestModels/Migrations/AnonymousArraySeedContext.cs
@@ -142,7 +142,7 @@ namespace {rootNamespace}.Migrations
                         .ValueGeneratedOnAdd()
                         .HasColumnType(""int"");
 
-                    b.Property<HierarchyId>(""HierarchyId"")
+                    b.Property<SqlHierarchyId?>(""HierarchyId"")
                         .HasColumnType(""hierarchyid"");
 
                     b.Property<string>(""Name"")

--- a/test/EFCore.SqlServer.HierarchyId.Tests/TestModels/Migrations/TypedArraySeedContext.cs
+++ b/test/EFCore.SqlServer.HierarchyId.Tests/TestModels/Migrations/TypedArraySeedContext.cs
@@ -144,7 +144,7 @@ namespace {rootNamespace}.Migrations
                         .ValueGeneratedOnAdd()
                         .HasColumnType(""int"");
 
-                    b.Property<HierarchyId>(""HierarchyId"")
+                    b.Property<SqlHierarchyId?>(""HierarchyId"")
                         .HasColumnType(""hierarchyid"");
 
                     b.Property<string>(""Name"")

